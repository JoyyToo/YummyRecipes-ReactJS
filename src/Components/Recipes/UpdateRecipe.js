import React, { Component }  from 'react';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import TextField  from 'material-ui/TextField';
import Paper from 'material-ui/Paper';
import { Link } from 'react-router-dom';
import {notify} from 'react-notify-toast'
import axiosInstance from '../Constants/AxiosCall';
import '../styles.css'


class UpdateRecipe extends Component {
    // initialize state
    state = {
        name: '',
        time: '',
        ingredients: '',
        procedure: '',
        id: '',
        category_id: '',
    }

    // handle user input
    handleInputChange = (event) => {
        const target = event.target;
        const value = target.value;
        const name = target.name;
        this.setState({[name]: value});
    };

    // handle get recipe request
    getRecipe() {
        const id = this.props.match.params['id'];
        const category_id = this.props.match.params['category_id'];
        const token = window.localStorage.getItem('token');
        if (!token) {
            window.location.replace('/login')
        }

        // send GET request to API
        axiosInstance.get(`category/${category_id}/recipes/${id}`)
            .then((response) => {
                let recipe = response.data['recipes'];
                this.setState({
                    id: id,
                    category_id: category_id,
                    name: recipe['name'],
                    time: recipe['time'],
                    ingredients: recipe['ingredients'],
                    procedure: recipe['procedure'],
                })
            })
            .catch((error) => {
                notify.show(error.response, 'error', 4000);
            });
    }

    // handle edit recipe request
    handleEditRecipe = () => {
        const payload = new FormData();
        payload.set('name', this.state.name)    
        payload.set('time', this.state.time,)
        payload.set('ingredients', this.state.ingredients,)
        payload.set('procedure', this.state.procedure,)

        // send PUT request to API
        axiosInstance.put(`category/${this.state.category_id}/recipes/${this.state.id}`, payload)
            .then((response) => {
                notify.show(response.data.message, 'success', 4000);
                this.props
                    .history
                    .push(`/categories/${this.state.category_id}/recipes`);
            })

            .catch((error) => {
                notify.show(error.response.data.message, 'error', 4000);
              
            });
    };

    // mount recipes and when page loads
    componentDidMount() {
        this.getRecipe();
    }
    
    // render edit recipe form
    render(){
        let style = {
            marginLeft: 20,
            width: 440,
          };
        let divstyle = {
            width: 500,
            margin: 'auto',
            textAlign: 'left',
            paddingLeft: 20,
        }
        return (
            <MuiThemeProvider>
            <div className="main">
                <div className="inputs" style={{paddingTop: '60px'}}>
                <Paper zDepth={2} style={divstyle}>
                    <p class="heading"><b>UPDATE RECIPE</b></p>

                    <TextField floatingLabelText="Name" name="name" 
                                 value={this.state.name} style={style} onChange={this.handleInputChange} /><br /><br />

                    <TextField floatingLabelText="Time taken" name="time" 
                                 value={this.state.time} style={style} onChange={this.handleInputChange} /><br /><br />

                    <TextField floatingLabelText="Ingridients" name="ingredients" 
                                 value={this.state.ingredients} style={style} multiLine={true} rows={3}
                                 onChange={this.handleInputChange} /><br />

                    <TextField floatingLabelText="Procedure" name="procedure" 
                                 value={this.state.procedure} style={style} multiLine={true} rows={3} 
                                 onChange={this.handleInputChange} /><br />

                    <div className="buttons" id="btnstyle">
                    <button className="network" id="one" onClick={(event => this.handleEditRecipe(event))}>UPDATE</button>
                    <Link to={"/categories/" + this.state.category_id + "/recipe/" + this.state.id}>
                    <button className="network" id="two">CANCEL</button></Link>

                    </div>

                    <p>A recipe has no soul. You are the cook, give life to the recipe</p> <br />

                </Paper><br />
                    
                </div>
            </div>
            </MuiThemeProvider>
        );
    }}

export default UpdateRecipe;
