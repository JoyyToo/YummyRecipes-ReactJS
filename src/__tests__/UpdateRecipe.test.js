import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import UpdateRecipe from '../Components/Recipes/UpdateRecipe'

describe('UpdateRecipe  component', () => {
    const params = {
        match: {
            params: {
                id:1
            }
        }
    }
    const wrapper = shallow(<UpdateRecipe  match={{params}}/>);
    it('renders properly', () => {
        expect(shallowToJson(wrapper)).toMatchSnapshot();
      });
  
    it('renders 1 <UpdateRecipe/> component', () =>{
        const component = shallow(<UpdateRecipe match={{params}}/>);
        expect(component).toHaveLength(1);
    });
  
    it('it returns props correctly', () =>{
        const component = shallow(<UpdateRecipe name="app" match={{params}}/>);
        expect(component.instance().props.name).toBe('app')
    });

    it('contains MuiThemeProvider', () => {
        expect(wrapper.find('MuiThemeProvider')).toHaveLength(1);
    });

    it('contains divs', () => {
        expect(wrapper.find('div')).toHaveLength(3);
    });

    it('contains Paper', () => {
        expect(wrapper.find('Paper')).toHaveLength(1);
    });

    it('contains p', () => {
        expect(wrapper.find('p')).toHaveLength(2);
    });

    it('contains TextField ', () => {
        expect(wrapper.find('TextField')).toHaveLength(4);
    });

    it('contains button', () => {
        expect(wrapper.find('button')).toHaveLength(2);
    });

    it('contains Link', () => {
        expect(wrapper.find('Link')).toHaveLength(1);
    });

    it('has initial state', () => {
        expect(wrapper.state().name).toEqual('');
        expect(wrapper.state().time).toEqual('');
        expect(wrapper.state().ingredients).toEqual('');
        expect(wrapper.state().procedure).toEqual('');
    });

    it('changes state', () => {
        wrapper.setState({ name: 'apple pie', time: '1 hour', ingredients: 'flour, pie', procedure:'bake' });
        expect(wrapper.find('[name="name"]').props().value).toEqual('apple pie');
        expect(wrapper.find('[name="time"]').props().value).toEqual('1 hour');
        expect(wrapper.find('[name="ingredients"]').props().value).toEqual('flour, pie');
        expect(wrapper.find('[name="procedure"]').props().value).toEqual('bake');
    });

    it('has the correct form fields', () => {
        expect(wrapper.find('[name="name"]')).toHaveLength(1);
        expect(wrapper.find('[name="time"]')).toHaveLength(1);
        expect(wrapper.find('[name="ingredients"]')).toHaveLength(1);
        expect(wrapper.find('[name="procedure"]')).toHaveLength(1);
    });

})