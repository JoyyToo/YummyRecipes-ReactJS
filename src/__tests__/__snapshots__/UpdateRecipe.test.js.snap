// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateRecipe  component renders properly 1`] = `
<MuiThemeProvider>
  <div
    className="main"
  >
    <div
      className="inputs"
      style={
        Object {
          "paddingTop": "60px",
        }
      }
    >
      <Paper
        circle={false}
        rounded={true}
        style={
          Object {
            "margin": "auto",
            "paddingLeft": 20,
            "textAlign": "left",
            "width": 500,
          }
        }
        transitionEnabled={true}
        zDepth={2}
      >
        <p
          className="heading"
        >
          <b>
            UPDATE RECIPE
          </b>
        </p>
        <TextField
          disabled={false}
          floatingLabelFixed={false}
          floatingLabelText="Name"
          fullWidth={false}
          multiLine={false}
          name="name"
          onChange={[Function]}
          rows={1}
          style={
            Object {
              "marginLeft": 20,
              "width": 440,
            }
          }
          type="text"
          underlineShow={true}
          value=""
        />
        <br />
        <br />
        <TextField
          disabled={false}
          floatingLabelFixed={false}
          floatingLabelText="Time taken"
          fullWidth={false}
          multiLine={false}
          name="time"
          onChange={[Function]}
          rows={1}
          style={
            Object {
              "marginLeft": 20,
              "width": 440,
            }
          }
          type="text"
          underlineShow={true}
          value=""
        />
        <br />
        <br />
        <TextField
          disabled={false}
          floatingLabelFixed={false}
          floatingLabelText="Ingridients"
          fullWidth={false}
          multiLine={true}
          name="ingredients"
          onChange={[Function]}
          rows={3}
          style={
            Object {
              "marginLeft": 20,
              "width": 440,
            }
          }
          type="text"
          underlineShow={true}
          value=""
        />
        <br />
        <TextField
          disabled={false}
          floatingLabelFixed={false}
          floatingLabelText="Procedure"
          fullWidth={false}
          multiLine={true}
          name="procedure"
          onChange={[Function]}
          rows={3}
          style={
            Object {
              "marginLeft": 20,
              "width": 440,
            }
          }
          type="text"
          underlineShow={true}
          value=""
        />
        <br />
        <div
          className="buttons"
          id="btnstyle"
        >
          <button
            className="network"
            id="one"
            onClick={[Function]}
          >
            UPDATE
          </button>
          <Link
            replace={false}
            to="/categories//recipe/\\""
          >
            <button
              className="network"
              id="two"
            >
              CANCEL
            </button>
          </Link>
        </div>
        <p>
          A recipe has no soul. You are the cook, give life to the recipe
        </p>
         
        <br />
      </Paper>
      <br />
    </div>
  </div>
</MuiThemeProvider>
`;
